src/lib/http2.ts

```diff
@@ -36,70 +31,45 @@ export interface Http2FetchResponse {
 export default class Http2Client {
   private _url: URL
   private _authority: string
-  private _defaultPath: string
   private _http2Opts: object
-  private _connected: ConnectState
-  private _connectPromise?: Promise<http2.ClientHttp2Session>
-  private _client?: http2.ClientHttp2Session
+  private _defaultPath: string
+  private _sessions: Array<Http2Session>
+  private _maxRequestsPerSession?: number
 
   constructor (url: string, opts?: object, http2Opts?: object) {
     this._url = new URL(url)
     this._authority = this._url.origin
     this._defaultPath = this._url.pathname
     this._http2Opts = http2Opts || {}
+    this._maxRequestsPerSession = opts && Number(opts['maxRequestsPerSession'])
 
-    this._connected = ConnectState.DISCONNECTED
+    this._sessions = []
   }
 
-  _connect (): Promise<http2.ClientHttp2Session> {
-    if (this._connected === ConnectState.CONNECTED && this._client) {
-      debug('client connected')
-      return Promise.resolve(this._client)
-    } else if (this._connected === ConnectState.CONNECTING && this._connectPromise) {
-      debug('client connecting')
-      return this._connectPromise
-    }
-
-    debug('creating client. authority=', this._authority)
-    const client = http2.connect(this._authority, this._http2Opts)
-
-    this._client = client
-    this._connected = ConnectState.CONNECTING
-    this._connectPromise = new Promise((resolve, reject) => {
-      const cleanUp = () => {
-        client.removeListener('connect', onConnect)
-        client.removeListener('error', onError)
-        client.removeListener('close', onClose)
-      }
-
-      const onConnect = () => {
-        debug('client connected.')
-        client.removeListener('connect', onConnect)
-        this._connected = ConnectState.CONNECTED
-        resolve(client)
-      }
-
-      const onClose = () => {
-        debug('client closed.')
-        cleanUp()
-        this._connected = ConnectState.DISCONNECTED
-        reject(new Error('closed while opening connection'))
-      }
-
-      // TODO: log the error
-      const onError = (error: Error) => {
-        debug('client encountered error. error=', error.message)
-        cleanUp()
-        this._connected = ConnectState.DISCONNECTED
-        reject(error)
+  async _allocateRequestAndRun<T>(cb: (client: http2.ClientHttp2Session) => Promise<T>): Promise<T> {
+    for (const session of this._sessions) {
+      const client = await session.allocateRequest()
+
+      if (client) {
+        try {
+          const result = await cb(client)
+          return result
+        } catch (e) {
+          throw e
```

Comment:
Why `catch` just to `throw`?

Reply:
:+1: didn't realize you could do try/finally instead of try/catch/finally